File: platforms/android/runtime/core/src/main/java/io/nativeblocks/runtime/api/provider/block/BlockProps.kt
package io.nativeblocks.runtime.api.provider.block

import androidx.compose.runtime.Composable
import androidx.compose.runtime.Immutable
import io.nativeblocks.runtime.api.provider.model.NativeActionModel
import io.nativeblocks.runtime.api.provider.model.NativeBlockModel
import io.nativeblocks.runtime.api.provider.model.NativeBlockSlotModel
import io.nativeblocks.runtime.api.provider.model.NativeVariableModel

const val NONE_INDEX = -1

internal typealias BlockComposable = @Composable (blockProps: BlockProps) -> Unit

/**
 * Represents the properties required for rendering and handling a native block.
 * @param instanceName instance name of NativeblocksManager.
 * @param listItemIndex Index of the list item associated with the block (optional).
 * @param onFindVariable Lambda function for retrieving a [NativeVariableModel] by its key.
 * @param onVariableChange Callback invoked when a variable changes (optional).
 * @param onFindAction Lambda function for retrieving a [NativeActionModel] by its event type.
 * @param onHandleAction Callback invoked to handle an action with the given index, action model, and type (optional).
 * @param block The model representing the block to be rendered (optional).
 * @param onSubBlock Composable callback to render a slot's sub-blocks; receives child block keys grouped by slot name, the slot, index, and scope (optional).
 */
@Immutable
data class BlockProps(
    val instanceName: String,
    val listItemIndex: Int,
    val onFindVariable: (String) -> NativeVariableModel?,
    val onVariableChange: (NativeVariableModel) -> Unit,
    val onFindAction: (String) -> NativeActionModel?,
    val onHandleAction: (Int, NativeActionModel?, String) -> Unit,
    val block: NativeBlockModel,
    val onSubBlock: @Composable (blockKeys: Map<String, List<String>>, slot: NativeBlockSlotModel, index: Int, scope: Any?) -> Unit
)

File: platforms/ios/runtime/Sources/NativeblocksRuntime/api/provider/block/BlockProps.swift
import Foundation
import SwiftUI

/// A constant representing an invalid or non-existent index.
public let NONE_INDEX = -1

/// Represents the properties associated with a native block.
/// The `BlockProps` struct is used to pass all the necessary information required to render a block, including variables, actions, and callbacks.
public struct BlockProps {
    /// Instance name of NativeblocksManager.
    public let instanceName: String

    /// The index of the item in the list that the block applies to (if applicable).
    public let listItemIndex: Int

    /// A function for retrieving a [NativeVariableModel] by its key.
    public let onFindVariable: (String) -> NativeVariableModel?

    /// Callback function to handle changes to a variable.
    public let onVariableChange: (NativeVariableModel) -> Void

    /// A function for retrieving a [NativeActionModel] by its event type.
    public let onFindAction: (String) -> NativeActionModel?

    /// Callback function to handle an action.
    /// - Parameters:
    ///   - listItemIndex: The index of the item in the list that the action applies to.
    ///   - action: The action model to be executed.
    ///   - type: The type of action being performed.
    public let onHandleAction: (Int, NativeActionModel?, String) -> Void

    /// The block model representing the current block.
    public let block: NativeBlockModel

    /// Callback function to render a slot's sub-blocks.
    /// - Parameters:
    ///   - blockKeys: Child block keys grouped by slot name.
    ///   - subSlot: The slot whose children should be rendered.
    ///   - itemIndex: The index of the item in the list for which sub-blocks are rendered.
    ///   - scope: Optional layout scope forwarded by the parent block.
    /// - Returns: An `AnyView` containing the rendered sub-blocks.
    public let onSubBlock: ([String: [String]], NativeBlockSlotModel, Int, Any?) -> AnyView
}
